lib/web/apiserver.go

```diff
@@ -2539,9 +2527,31 @@ func (h *Handler) clusterUnifiedResourcesGet(w http.ResponseWriter, r *http.Requ
 			}
 			unifiedResources = append(unifiedResources, server)
 		case types.DatabaseServer:
+			if !containsDb {
+				dbNames, dbUsers, err = getDatabaseUsersAndNames(accessChecker)
+				if err != nil {
+					return nil, trace.Wrap(err)
+				}
+				containsDb = true
+			}
 			db := ui.MakeDatabase(r.GetDatabase(), dbUsers, dbNames)
 			unifiedResources = append(unifiedResources, db)
 		case types.AppServer:
+			if !containsApp {
+				userGroups, err = apiclient.GetAllResources[types.UserGroup](request.Context(), clt, &proto.ListResourcesRequest{
+					ResourceType:     types.KindUserGroup,
+					Namespace:        apidefaults.Namespace,
+					UseSearchAsRoles: true,
+				})
+				if err != nil {
+					h.log.Debugf("Unable to fetch user groups while listing applications, unable to display associated user groups: %v", err)
+				}
+
+				for _, userGroup := range userGroups {
+					userGroupLookup[userGroup.GetName()] = userGroup
+				}
+				containsApp = true
+			}
```

Comment:
nit: can we make this reusable as an anonymous function `getUserGroupLookup` rather than duplicating the logic?

suggestion: we could make it generator-like so the `!containsApp` logic is contained within the anonymous function:

```go
// generator to retrieve UserGroupLookup on first call and return it again in subsequent calls.
// If we encounter an error, we log it once and return an empty UserGroupLookup for the current and subsequent calls.
getUserGroupLookup := func() func() (map[string]types.UserGroup) {
	userGroupLookup := make(map[string]types.UserGroup)
	var gotUserGroupLookup bool
	return func() map[string]types.UserGroup {
		if gotUserGroupLookup {
			return userGroupLookup
		}
		
		userGroups, err = apiclient.GetAllResources[types.UserGroup](request.Context(), clt, &proto.ListResourcesRequest{
			ResourceType:     types.KindUserGroup,
			Namespace:        apidefaults.Namespace,
			UseSearchAsRoles: true,
		})
		if err != nil {
			h.log.Debugf("Unable to fetch user groups while listing applications, unable to display associated user groups: %v", err)
		}

		for _, userGroup := range userGroups {
			userGroupLookup[userGroup.GetName()] = userGroup
		}

		gotUserGroupLookup = true
		return userGroupLookup
	}
}()

...

app := ui.MakeApp(r.GetApp(), ui.MakeAppsConfig{
	LocalClusterName:  h.auth.clusterName,
	LocalProxyDNSName: h.proxyDNSName(),
	AppClusterName:    site.GetName(),
	Identity:          identity,
	UserGroupLookup:   getUserGroupLookup(),
	Logger:            h.log,
})
```

Reply:
absolutely, this look awesome. thank you! will test and add